File: cpp/questions/MedianFinder.cpp
#include <bits/stdc++.h>

using namespace std;

class MedianFinder {
private:
    std::priority_queue<int, std::vector<int>, std::greater<>> gq;
    std::priority_queue<int, std::vector<int>, std::less<>> lq;

public:
    MedianFinder() {}
    
    void addNum(int num) {
        if (lq.size() <= gq.size()) {
            lq.push(num);
        } else {
            gq.push(num);
        }
        if (lq.empty() || gq.empty()) {
            return;
        }
        while (lq.top() > gq.top()) {
            int lt = lq.top();
            int gt = gq.top();
            lq.pop();
            gq.pop();
            lq.push(gt);
            gq.push(lt);
        }
    }
    
    double findMedian() {
        if (lq.size() == gq.size()) {
            return (lq.top()+gq.top()) / 2.0;
        } else {
            return static_cast<double>(lq.top());
        }
    }
};

File: python/questions/MedianFinder.py
import heapq

class MedianFinder:

    def __init__(self):
        self.gq = []
        self.lq = []

    def addNum(self, num: int) -> None:
        if len(self.lq) <= len(self.gq):
            heapq.heappush(self.lq, -num)
        else:
            heapq.heappush(self.gq, num)
        if not len(self.lq) or not len(self.gq):
            return
        while -self.lq[0] > self.gq[0]:
            lt = -heapq.heappop(self.lq)
            gt = heapq.heappop(self.gq)
            heapq.heappush(self.lq, -gt)
            heapq.heappush(self.gq, lt)

    def findMedian(self) -> float:
        if len(self.lq) == len(self.gq):
            return (self.gq[0] - self.lq[0]) / 2
        else:
            return -self.lq[0]

